listener/avalon_listener/tcs_work_order_handler.py

```diff
@@ -234,13 +260,18 @@ def WorkOrderSubmit(self, **params):
 
         if((self.workorder_count + 1) > self.max_workorder_count):
 
+            # wo_ids is a csv of work order ids retrieved from database
+            wo_ids = self.kv_helper.get("wo-worker-processed", worker_id)
+            processed_wo_ids = [] if wo_ids is None else wo_ids.split(",")
+
             # if max count reached clear a processed entry
             work_orders = self.kv_helper.lookup("wo-timestamps")
             for id in work_orders:
 
                 # If work order is processed then remove from table
-                if(self.kv_helper.get("wo-processed", id) is not None):
-                    self.kv_helper.remove("wo-processed", id)
+                if id in processed_wo_ids:
+                    self.kv_helper.csv_search_delete(
```

Comment:
csv_search_delete() is called for every work order request and It is O(n) operation, we might come across bottleneck here because all other workers are blocked until this completed.

Reply:
We will hit this condition only when work orders in DB is more than max work order count. So this won't be called for every work order request. For clarity, it's better to move this block to a separate function. Also (different PR) we could start a different thread for cleanup of work orders instead of doing it as part of workOrderSubmit handling

Reply:
True @Ram-srini, that it will be O(n) and delay will be directly proportional to the number of work order id that would have been processed. But it might not be called for all work order. Rather should not be. Ii would only be so when the queue is almost topping out and that in turn would happen at the maximum possible throughput that can be churned out of the system. Having said that, still the performance at that point of time should also matter. this was chosen over introducing another table to ensure operations are as atomic as possible and refrain from multiple updates to reflect same state change.

On a side note, since its an interesting topic, I will continue on to say that the average time complexity for the operation would be in fact close to O(1). This is because a search is done on a queue which already maintains the order same as the one maintained over the items being iterated upon here. So, there is a very high chance(~99%) that the first item is hit every time in the search.